File: Java/0401-0450/446. Arithmetic Slices II - Subsequence.java
// https://leetcode.com/problems/arithmetic-slices-ii-subsequence/

// https://leetcode.com/problems/arithmetic-slices-ii-subsequence/solutions/4521112/o-n-2-java-python-javascript-dynamic-programming-hashmap/?envType=daily-question&envId=2024-01-07

class Solution {
    public int numberOfArithmeticSlices(int[] nums) {
        HashMap<Long, List<Integer>> storage = new HashMap<>();
        int size = nums.length;
        int[][] dp = new int[size][size];

        int result = 0;
        for(int index=0; index<size; index++){
            long num = nums[index];
            storage.computeIfAbsent(num, k -> new ArrayList<>()).add(index);}

        for(int row=1; row<size; row++){
            for(int col=row+1; col<size; col++){
                long prev = 2L * nums[row] - nums[col];
                if (prev > Integer.MAX_VALUE || prev < Integer.MIN_VALUE)
                    continue;
                if(storage.containsKey(prev) ){
                    for(int k : storage.get(prev)){
                        if(k < row){
                            dp[row][col] += dp[k][row] + 1;
                        }else{
                            break;
                        }
                    }
                }
                result += dp[row][col];
            }
        }

        return result;
    }
}

File: Python/0401-0450/446. Arithmetic Slices II - Subsequence.py
# https://leetcode.com/problems/arithmetic-slices-ii-subsequence/

# https://leetcode.com/problems/arithmetic-slices-ii-subsequence/solutions/4521112/o-n-2-java-python-javascript-dynamic-programming-hashmap/?envType=daily-question&envId=2024-01-07

class Solution:
    def numberOfArithmeticSlices(self, nums):
        storage = {}
        size = len(nums)
        dp = [[0] * size for _ in range(size)]

        for index in range(size):
            if nums[index] not in storage:
                storage[nums[index]] = []
            storage[nums[index]].append(index)

        result = 0
        for row in range(1, size):
            for col in range(row + 1, size):
                prev = 2 * nums[row] - nums[col]
                if prev in storage:
                    for k in storage[prev]:
                        if k < row:
                            dp[row][col] += dp[k][row] + 1
                        else:
                            break
                result += dp[row][col]

        return result
